src/tropicalnumber.jl

```diff
@@ -0,0 +1,70 @@
+const pm_TropicalNumber_suppAddition = Union{pm_Min, pm_Max}
+const pm_TropicalNumber_suppScalar = Union{pm_Rational}
+
+#convert input to supported Scalar type
+pm_TropicalNumber{A,S}(scalar::Real) where {A <: pm_TropicalNumber_suppAddition, S <: pm_TropicalNumber_suppScalar} = pm_TropicalNumber{A,S}(convert(S,scalar))
+
+#pm_Rational es default Scalar
+pm_TropicalNumber{A}(scalar::Real) where A <: pm_TropicalNumber_suppAddition = pm_TropicalNumber{A,pm_Rational}(scalar)
+pm_TropicalNumber{A}() where A <: pm_TropicalNumber_suppAddition = pm_TropicalNumber{A,pm_Rational}()
+
+function Base.:(==)(x::pm_TropicalNumber{A,S},y::Real) where {A <: pm_TropicalNumber_suppAddition, S <: pm_TropicalNumber_suppScalar}
+    return x == pm_TropicalNumber{A,S}(y)
+end
+
+Base.:(==)(x::Real,y::pm_TropicalNumber{A,S}) where {A <: pm_TropicalNumber_suppAddition, S <: pm_TropicalNumber_suppScalar} = (y == x)
+
+#catch wrong typings for basic operations
+function Base.:+(x::pm_TropicalNumber{A1,S1},y::pm_TropicalNumber{A2,S2}) where {A1,A2 <: pm_TropicalNumber_suppAddition, S1,S2 <: pm_TropicalNumber_suppScalar}
```

Comment:
btw. if we're not promoting anything then this could be shortened to
```julia
Base.promote_rule(::Type{<:pm_TropicalNumber{A, S}}, ::Type{<:pm_TropicalNumber{A, T}}) = pm_TropicalNumber{A, promote_type(S,T)}
for op in (:+, :*, ://, :/, :<)
    @eval begin
        Base.$(op)(x::pm_TropicalNumber{A, S}, y::pm_TropicalNumber{A, T}) where {A,S,T} = $(op)(promote(x,y)...)
        Base.$(op)(x::pm_TropicalNumber{A}, y::pm_TropicalNumber{B}) where {A,B} = 
            throw(DomainError((x,y), "The operation of $(op) for tropical numbers with $A$ and $B$ is not defined"))
    end
end
```
* At the moment the promote rule and the first definition of `Base.$(op)` will never be called as we only allow `pm_Rational` as the second type, but we can cheaply have them for future
* we throw error based on the mismatch of the addition types

Reply:
unluckily, this version (and any modification of it i have thought of) gives us an `UndefVarError`:

> ERROR: UndefVarError: op not defined

any idea how to fix this?

Reply:
forgot about escaping ;-) this should work:
```julia
for op in (:+, :*, ://, :/, :<)
    @eval begin
        Base.$(op)(x::pm_TropicalNumber{A, S}, y::pm_TropicalNumber{A, T}) where {A,S,T} = $(op)(promote(x,y)...)
        Base.$(op)(x::pm_TropicalNumber{A}, y::pm_TropicalNumber{B}) where {A,B} = 
            throw(DomainError((x,y), "The operation of $(esc(op)) for tropical numbers with $(esc(A)) and $(esc(B)) is not defined"))
    end
end
```

Reply:
it still does not work. i should add that the line that is referenced to is not the one you altered but the one starting with `Base.$(op)`

Reply:
sorry, interpolation is hard ;) Try:
```julia
julia> struct pm_TropicalNumber{A, S} <: Number
          val::S
       end

julia> struct pm_Min end

julia> struct pm_Max end

julia> pm_TropicalNumber{A}(x::S) where {A,S} = pm_TropicalNumber{A, S}(x)

julia> for op in (:+, :*, ://, :/, :<)
           @eval begin
               Base.$(op)(x::pm_TropicalNumber{A, S}, y::pm_TropicalNumber{A, T}) where {A,S,T} = $(op)(promote(x,y)...)
               Base.$(op)(x::pm_TropicalNumber{A}, y::pm_TropicalNumber{B}) where {A,B} = 
                   throw(DomainError((x,y), "The operation $(string($op)) for tropical numbers with $A and $B is not defined"))
           end
       end

julia> t = pm_TropicalNumber{pm_Min}(2//1); s = pm_TropicalNumber{pm_Max}(2.0);

julia> t+s
ERROR: DomainError with (pm_TropicalNumber{pm_Min,Rational{Int64}}(2//1), pm_TropicalNumber{pm_Max,Float64}(2.0)):
The operation + for tropical numbers with pm_Min and pm_Max is not defined
Stacktrace:
 [1] +(::pm_TropicalNumber{pm_Min,Rational{Int64}}, ::pm_TropicalNumber{pm_Max,Float64}) at ./REPL[26]:4
 [2] top-level scope at REPL[28]:1

julia> t*s
ERROR: DomainError with (pm_TropicalNumber{pm_Min,Rational{Int64}}(2//1), pm_TropicalNumber{pm_Max,Float64}(2.0)):
The operation * for tropical numbers with pm_Min and pm_Max is not defined
Stacktrace:
 [1] *(::pm_TropicalNumber{pm_Min,Rational{Int64}}, ::pm_TropicalNumber{pm_Max,Float64}) at ./REPL[26]:4
 [2] top-level scope at REPL[29]:1

```